File: Reboot2025/SlidingWindow/LongestSubstringKUniqueCharacters.py
def LongestSubstringkUniqueCharacters(string: str, k: int):
    """
    Calculate the length of the longest substring with at most k unique characters.

    Parameters:
    string (str): The input string
    k (int): The maximum number of unique characters allowed in the substring

    Returns:
    int: The length of the longest substring with at most k unique characters
    """
    i = 0
    max_len = float("-inf")
    hashmap = {}
    for j, item in enumerate(string):
        if item in hashmap:
            hashmap[item] += 1
        else:
            hashmap[item] = 1

        if len(hashmap) < k:
            continue
        if len(hashmap) == k:
            max_len = max(max_len, j - i + 1)
        else:
            while len(hashmap) > k:
                if string[i] in hashmap:
                    hashmap[string[i]] -= 1
                    if hashmap[string[i]] == 0:
                        del hashmap[string[i]]
                i += 1
            if len(hashmap) == k:
                max_len = max(max_len, j - i + 1)
    return max_len


S = "aabacbebebe"
K = 3

print(LongestSubstringkUniqueCharacters(S, K))

File: Reboot2025/SlidingWindow/LongestSubstringKUniqueCharacters.java
package Reboot.SlidingWindow;

import java.util.HashMap;

public class LongestSubstringKUniqueCharacters {
    public static void main(String[] args) {
        String s = "aabacbebebe";
        int k = 3;
        System.out.println(longestSubstringkUniqueCharacters(s, k));
    }
    public static int longestSubstringkUniqueCharacters(String s, int k) {
        int i = 0;
        int maxLen = Integer.MIN_VALUE;
        HashMap<Character, Integer> hashMap = new HashMap<>();
        for (int j = 0; j < s.length(); j++) {
            Character c = s.charAt(j);
            if (hashMap.containsKey(s.charAt(j))) {
                hashMap.put(c, hashMap.get(c) + 1);
            } else {
                hashMap.put(c, 1);
            }
            if (hashMap.size() < k) {
                continue;
            } else if (hashMap.size() == k) {
                maxLen = Math.max(maxLen, j - i + 1);
            } else {
                while (hashMap.size() > k) {
                    if (hashMap.containsKey(s.charAt(i))) {
                        hashMap.put(s.charAt(i), hashMap.get(s.charAt(i)) - 1);
                        if (hashMap.get(s.charAt(i)) == 0) {
                            hashMap.remove(s.charAt(i));
                        }
                    }
                    i += 1;
                }
            }
        }
        return maxLen;
    }
}
